### PR TITLE
Code Adventures & IO Update to New API

### PR DESCRIPTION
**Pull request by jamesm2w**
### Fluid Sim: 
- Moved to `src/examples/`
- Updated to take advantage of new API methods/kernels
- Didn't use new properties for performance at the moment, could be done in future.

### Makefile: 
- Simplified source variables with new layout for code under `src/`

### Including IO Changes: 
- Move IO under `src/`, 
- Rename to remove `mock_`
- Introduce IO Manager which controls when to write out to a file
- update ASCII and HDF5 writers to use new API methods

### Including API Changes:
- Add initialisation of Particle forces in constructor. If force maps are uninitialised it causes all sorts of errors.
- Addition of some consequential getter functions like `get_id` which are useful
- Changed vector magnitude to actually return the magnitude not the squared sum ( & consequential updates )
- Reintroduced defs for Vec2D and Vec3D
- Reincluded stream output for vectors, fixed display as well
- Add IO output on each iteration of the simulation running
- Reintroduce the extra functions (string_format and sgn)

### Testing Changes:
- Updated IO Test to use new API functions
- Updated Vector Test to cover more vector operations for the fun of it

### CODE REVIEW

File: src/wash/vector.hpp [hunk 1]
```diff
@@ -119,4 +123,18 @@ namespace wash {
     };
 
     using SimulationVecT = Vec<double, DIM>;
+    
+    typedef Vec<double, 2> Vec2D;
+    typedef Vec<double, 3> Vec3D;
 }
+
+template <typename T, int dim>
+std::ostream& operator<<(std::ostream& s, const wash::Vec<T, dim>& vec) {
```
Comment (mwasacz) [inline src/wash/vector.hpp:132]: Is there a reason why this is outside the `wash` namespace?

File: src/wash/wash.cpp [hunk 2]
```diff
@@ -93,13 +95,22 @@ namespace wash {
...
+
+        io.handle_iteration(-1);
+        
         for (uint64_t iter = 0; iter < max_iterations; iter++) {
             for (auto& k : loop_kernels) {
                 k->exec();
             }
+
+            io.handle_iteration(iter);
+            std::cout << "Finished iter " << iter << std::endl;
```
Comment (mwasacz) [inline src/wash/wash.cpp:113]: Not sure if printing debug info here is the best idea. Maybe surround it with `#ifdef DEBUG`?

File: src/wash/particle.cpp [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#include "wash.hpp"
+
+namespace wash {
+    void Particle::initialise_particle_forces() {
```
Comment (mwasacz) [inline src/wash/particle.cpp:4]: Looks like I forgot to copy this function from the old API, thanks for fixing this

**Review by mwasacz — COMMENTED** [review]
Looks good overall, just a few minor comments

File: src/wash/vector.hpp (hunk 1 above)
Comment (jamesm2w) [reply src/wash/vector.hpp:132]: This is just providing an override for the generic `<<` operator for the `wash::Vec` type. I don't think it makes sense for it to be namespaced

File: src/wash/wash.cpp (hunk 2 above)
Comment (jamesm2w) [reply src/wash/wash.cpp:113]: I've found when running long-running programs especially over the batch-compute machines it's useful to have some output to just show it making progress. Not sure if it's really "debugging" in that sense.

File: src/wash/wash.hpp [hunk 4]
```diff
@@ -198,4 +201,35 @@ namespace wash {
...
+    /**
+     * @brief Helper function to format to a c++ string
+     * 
+     * @tparam Args 
+     * @param format standard style format string
+     * @param args arguments for the format string
+     * @return std::string 
+     */
+    template <typename... Args>
+    std::string string_format(const std::string& format, Args... args) {
```
Comment (Scott943) [inline src/wash/wash.hpp:214]: Could we keep these helper functions in a separate file like the previous `wash_extra_solutions.cpp`?

**Review by Scott943 — COMMENTED** [review]
Nice work. 

I think we need to figure out how to incoroporate some more automatic tests for the wash namespace and the code_adventures library, but I think that may be difficult

File: src/wash/wash.cpp (hunk 2 above)
Comment (mwasacz) [reply src/wash/wash.cpp:113]: Okay, fair point

File: src/wash/wash.hpp (hunk 4 above)
Comment (jamesm2w) [reply src/wash/wash.hpp]: Sure. Moved to a separate file.

**Pull request merged**